ENIGMAsystem/SHELL/Universal_System/Resources/AssetArray.h

```diff
@@ -41,8 +41,71 @@
 
 namespace enigma {
 
+template<typename T, int LEFT> 
+class OffsetVector {
+  std::vector<T> data_owner_;
+  T* data_;
+
+ public:
+  OffsetVector(): data_(nullptr) {}
+  OffsetVector(const OffsetVector<T, LEFT> &other):
+    data_owner_(other.data_owner_), data_(data_owner_.data() - LEFT) {}
+  size_t size() const {
+    return data_owner_.size() + LEFT;
+  }
+  T *data() { return data_; }
+  const T *data() const { return data_; }
+  template<typename... U> size_t push_back(U... args) {
+    data_owner_.push_back(args...);
+    data_ = data_owner_.data() - LEFT;
+    return size() - 1;
+  }
+  template<typename... U> size_t emplace_back(U... args) {
+    data_owner_.emplace_back(args...);
+    data_ = data_owner_.data() - LEFT;
+    return size() - 1;
+  }
+  T& operator[](int index) {

```

Comment:
Technically better to do these as `template<typename ind_t> T &operator[](ind_t index)`.